### PR TITLE
[SPARK-22217] [SQL] ParquetFileFormat to support arbitrary OutputCommitters

### PR DESCRIPTION
**Pull request by steveloughran**
## What changes were proposed in this pull request?

`ParquetFileFormat` to relax its requirement of output committer class from `org.apache.parquet.hadoop.ParquetOutputCommitter` or subclass thereof (and so implicitly Hadoop `FileOutputCommitter`) to any committer implementing `org.apache.hadoop.mapreduce.OutputCommitter`

This enables output committers which don't write to the filesystem the way `FileOutputCommitter` does to save parquet data from a dataframe: at present you cannot do this.

Before a committer which isn't a subclass of `ParquetOutputCommitter`, it checks to see if the context has requested summary metadata by setting `parquet.enable.summary-metadata`. If true, and the committer class isn't a parquet committer, it raises a RuntimeException with an error message.

(It could downgrade, of course, but raising an exception makes it clear there won't be an summary. It also makes the behaviour testable.)

Note that `SQLConf` already states that any `OutputCommitter` can be used, but that typically it's a subclass of ParquetOutputCommitter. That's not currently true. This patch will make the code consistent with the docs, adding tests to verify,

## How was this patch tested?

The patch includes a test suite, `ParquetCommitterSuite`, with a new committer, `MarkingFileOutputCommitter` which extends `FileOutputCommitter` and writes a marker file in the destination directory. The presence of the marker file can be used to verify the new committer was used. The tests then try the combinations of Parquet committer summary/no-summary and marking committer summary/no-summary. 

| committer | summary | outcome |
|-----------|---------|---------|
| parquet   | true    | success |
| parquet   | false   | success |
| marking   | false   | success with marker |
| marking   | true    | exception |

All tests are happy.

### CODE REVIEW

**Comment by steveloughran** [comment]
+ @rdblue

**Comment by rdblue** [comment]
+1

I completely agree that using a ParquetOutputCommitter should be optional.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (vanzin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: `IllegalArgumentException` or some other better exception?

**Review by vanzin — COMMENTED** [review]
One minor suggestion otherwise LGTM.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: How about `require` maybe?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Looks we can remove this `s` BTW.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala]: maybe `super.afterAll()`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala]: I think It might be a little bit better to use named arguments for readability: `writeDataFrame(MarkingFileOutput.COMMITTER, summary = false, check  = true)`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala]: Maybe just`val committer`.

**Review by HyukjinKwon — APPROVED** [review]
LGTM too, just few tiny nits while double checking.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: will do

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Depends on the policy about "what to do if it's not a parquet committer *and* the option for  job summaries is set. It could just mean "you don't get summaries", which worksforme :). May want to log at info though?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala
Comment (steveloughran) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala]: OK

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Oh, I mean .. s in `s" .. "`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala
Comment (steveloughran) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala]: done, + will add a check for spark==null so if a failure happens during setup, the exception doesn't get lost in teardown

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: aah

Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: aah. in the move to require() everything is going back onto a single line. so now moot

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: We need to issue an `AnalysisException` here.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala]:
```Scala
    try {
      ...
    } finally {
      super.afterAll()
    }
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: `AnalysisException`? Shouldn't this be `SparkException`? By the time this runs, Spark has already analyzed, optimized, and planned the job. Doesn't seem like failing analysis is appropriate.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: `SparkException` is better. Normally, we want to issue a Spark-specific exception type.

Comment (vanzin) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: `SparkException` makes it sound like it's a problem that Spark caused in some way. While this is caused by user input being incorrect, in which case the suggested `IllegalArgumentException` (which `require` throws) is better imo.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: In Spark SQL, we do issue the `AnalysisException` in many similar cases. I am also fine to use `SparkException`. 

In this specific case, the users are able to control the conf to make it works. Thus, we also need to improve the message to let users know how to resolve it by changing the conf.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala
Comment (steveloughran) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala]: good point

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: There's another option which is "log @ warn and continue". If someone has changed the committer, they get the consequences. That could also permit someone with a modified committer to generate schema summaries if they chose/permitted.

IT'd simplify this patch, need the tests tweaked...I'd change the SQLConf text with the committer key to say "if the committer isn't a ParquetOutputCommitter then don't expect summaries"

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala
Comment (jiangxb1987) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetCommitterSuite.scala]: nit: `commtters` -> `committers`

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Comment by jiangxb1987** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: I think I'd prefer the warn & continue option. It does little good to fail so late in a job, when the caller has already indicated that they want to use a different committer. Let them write the data out since this isn't a correctness issue, and they can add a summary file later if they want. Basically, there's less annoyance and interruption by not writing a summary file than by failing a job and forcing the user to re-run near the end.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: nit:

```scala
  ...
  s"Committer $committerClass is not a ParquetOutputCommitter and cannot create job summaries. " +
  s"Set Parquet option '${ParquetOutputFormat.ENABLE_JOB_SUMMARY}' to false.")
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: +1 for warn and continue.

**Comment by HyukjinKwon** [comment]
Still LGTM except for few nits.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: I'd thought about that; didn't look any better or worse. Will change it for log message.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: If we issuing a warning log, we will see such a warning message for each write operation. Does it look annoying?

**Comment by gatorsmile** [comment]
LGTM pending Jenkins

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: yes. there is that. Options: do something complicated with a static field to only print ones. Log at debug so people only see the message if they are trying to track things down.

Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: I think once per write operation is fine. It's not like it is once per file.

**Comment by rdblue** [comment]
Still +1 from me as well.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Hi, All.
Can we have this in Apache Spark 2.2.1?

**Comment by HyukjinKwon** [comment]
I didn't backported this one respecting the JIRA issue type, `Improvement` but yea, it sounds more like a bug fix.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: D'oh, `s` ...

**Comment by gatorsmile** [comment]
This is not eligible for backporting. We should not do it next time.

**Comment by HyukjinKwon** [comment]
I think this is a bug to fix as the previous behaviour does not work as documented:

```
subclass of org.apache.hadoop.mapreduce.OutputCommitter...
```

and does not change existing behaviour.

Could you elaborate why you think it is not eligible?

**Comment by gatorsmile** [comment]
That conf is an internal one. The end users will not see it. This is not a bug fix. 

We should not extend the existing functions or introduce new behaviors/features in 2.2.x releases.

**Comment by gatorsmile** [comment]
Since the risk is low, I did not revert it.

**Comment by HyukjinKwon** [comment]
How come fixing the behaviour as documented is not a bug fix? I think that basically mean we don't backport fixes for things not working as documented for other internal configurations.

This does not extend the functionailities. This fixes functionalities to work as documented and expected, and I call it a bugfix.

**Comment by gatorsmile** [comment]
This one starts at least since Spark 1.5. If you are not confident whether this is bug or not, please check it before merging it.

**Comment by HyukjinKwon** [comment]
I did this as I was confident if it is a bug because doc says it should work but actually not, without breaking the previous support.

**Comment by gatorsmile** [comment]
Ok. Next time, please check it with the committers who are familiar with Spark SQL.

**Comment by HyukjinKwon** [comment]
Will check it if I am not confident next time.

**Comment by steveloughran** [comment]
Thanks for reviewing this/getting it in. Personally, I had it in the "improvement" category rather than bug fix. If it wasn't for that line in the docs, there'd be no ambiguity about improve/vs fix, and there is always a lower-risk way to fix doc/code mismatch: change the docs.

But I'm grateful for it being in; with the backport to branch-2 ryan should be able to use it semi-immediately

**Comment by steveloughran** [comment]
PS, for people who are interested in dynamic committers, [MAPREDUCE-6823](https://issues.apache.org/jira/browse/MAPREDUCE-6823) is something to look at. It allows you to switch committers under pretty much everything *other than parquet*...this patch helps make Parquet manageable too

**Comment by HyukjinKwon** [comment]
I guess we wouldn't change the docs in branch-2.2 alone as we have a safe fix here for this mismatch anyway. I think I just wanted to say this backport can be justified.

**Comment by gatorsmile** [comment]
@steveloughran Thanks for your inputs. Totally agree on your opinions. 

Spark is an infrastructure software. We have to be very careful when backporting the PRs.

**Comment by yhuai** [comment]
@HyukjinKwon branch-2.2 is in a maintenance branch, I am not sure it is appropriate to merge this change to branch-2.2 since it is not really a bug fix. If the doc is not accurate, we should fix the doc. For a maintenance branch, we need to be very careful on what we merge and we should always avoid of unnecessary changes.

**Comment by HyukjinKwon** [comment]
Okay. I am sorry for this trouble. Should we revert this if you guys feel strongly about it?

**Comment by rdblue** [comment]
I have a lot of sympathy for the argument that infrastructure software shouldn't have too many backports and that those should be generally bug fixes. But, if I were working on a Spark distribution at a vendor, this is something I would definitely include because it's such a useful feature. I think that by not backporting this, we're just pushing that work downstream. Plus, the risk to adding this is low: the main behavior change is that users can specify a previously-banned committer for Parquet writes. Is it a bug fix? Probably not. But it fixes a big blocker.

**Comment by yhuai** [comment]
I am not really worried about this particular change. It's already merged and it seems a small and safe change. I am not planning to revert it.

But, in general, let's avoid of merging changes that are not bug fixes to a maintenance branch. If there is an exception, it will be better to make it clear earlier.

**Comment by HyukjinKwon** [comment]
Sure, I will and let me note it ahead next time. I made a mistake while trying to think of reasons for this backport.

**Comment by yhuai** [comment]
Thank you :)

**Comment by steveloughran** [comment]
> But, if I were working on a Spark distribution at a vendor, this is something I would definitely include because it's such a useful feature.

I concur :)

**Pull request closed without merging**
